### PR TITLE
Implement conditional pick

### LINKED ISSUE #3: [Feature Request] Filter Function in Queries
```
const target1 = {
  author: 'Leeingnyo',
  uri: 'https://github.com/Leeingnyo/pick-recursively',
  country: 'Korea'
};
const query1 = {
  author: true,
  uri: (uri) => uri.indexOf('https') == 0, // objects that contains https uri would be picked
  nothing: true
};
pick(target1, query1);
```

I think use filter functions in queries would be useful. (Especially in arrays)

How do you think about this?

### PR DESCRIPTION
**Pull request by Leeingnyo**
I implement the feature on #3 
This pull request will resolve #3

### CODE REVIEW

**Comment by HurSungYun** [comment]
```
assert.deepEqual(pick(target3, {
  articles: {
    article: article => {
      if (article.content.indexOf('2') < 0) {
        return true;
      }
      return false; // drop out OBJECT which contains '2' in content
    }
  }
}), {
  articles: [
    { title: 'title 1',
      content: 'content 1', },
      // content 2 is dropped out
    { title: 'title 3',
      content: 'content 3', }
  ]
});
```

I thought filtering would be like examples above.

I'm not sure filtering PROPERTY have benefit rather than filtering OBJECT.

**Comment by HurSungYun** [comment]
If you have reasons for these interface, it's okay.

But I don't know how to manipulate this object. (because different properties have been picked)

```
var a = [
    {
      isSecret: false,
      title: 'title 1'
    },
    {
      isSecret: true
    },
    {
      isSecret: false,
      title: 'title 3'
    }
  ]
```

I think result object below make sense and more useful.

```
var b = [
    {
      isSecret: false,
      title: 'title 1'
    },
    {
      isSecret: false,
      title: 'title 3'
    }
  ]
```

**Comment by HurSungYun** [comment]
I believed pick module can be used like SQL queries.

And it may be my mistake for requesting vague feature. I should have showed you an example. sorry for that.

Anyway, it's all up to you concluding these issues :)

**Comment by HurSungYun** [comment]
https://stackoverflow.com/questions/40421825/how-to-use-object-values-on-server-side-in-node-js

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_objects/Object/values

And Object.values() are in ES2017, not ES6.

so you have to decide not to use Object.values() or change ES version for pick-recursively

or use like this

```
objValues = Object.values(obj),
  // Object.keys() and map(),
  objKeysMap = Object.keys(obj).map((k) => obj[k]);
```

File: index.js [hunk 1]
```diff
@@ -11,10 +21,20 @@ module.exports = function pick(target, query) {
...
-        ) ? { [key]: pick(target[key], query[key]) } : {}), {});
+        .reduce((result, key) => {
+          let innerTarget = target[key], innerQuery = query[key];
+          let picked = {};
+          if (innerQuery === true || typeof innerQuery === 'object' && innerQuery !== null ||
+              typeof innerQuery === 'string') {
+            picked = { [key]: pick(innerTarget, innerQuery) };
+          } else if (typeof innerQuery === 'function') {
+            let evaluated = innerQuery(clone(innerTarget));
+            if (evaluated !== false) {
```
Comment (HurSungYun) [inline index.js:32]: If you want to filter OBJECT,

```
let evaluated = innerQuery(clone(innerTarget));
if (evaluated == false) {
   return null;
} else {
   picked = { [key]: pick(innerTarget, evaluated) };
}
```

and other things should be changed also.

**Comment by Leeingnyo** [comment]
You can use filter functions as you want.

```
const target = [
  {
    isSecret: false,
    title: 'title 1',
    content: 'content 1'
  },
  {
    isSecret: true,
    title: 'title 2',
    content: 'content 2'
  },
  {
    isSecret: false,
    title: 'title 3',
    content: 'content 3'
  }
]
```

```
const myQuery = { article: article => !article.isSecret ? ['isSecret', 'title'] : 'isSecret' };
const yourQuery = { article: article => !article.isSecret ? ['isSecret', 'title'] : false };
```

I used `myQuery ` at the example because I wanted to show that the return value of a filter function can have various type of query (here, string of array and string).
If you want the result like `b` on your reply (https://github.com/Leeingnyo/pick-recursively/pull/5#issuecomment-331696183), you should use `yourQuery `. The return value `false` makes the element drop out.

image

---

I'll use `myQuery` at this situation.

```
var articlesForList = pick(target, myQuery)
/*
result is
  [
    {
      isSecret: false,
      title: 'title 1'
    },
    {
      isSecret: true
    },
    {
      isSecret: false,
      title: 'title 3'
    }
  ]
*/
artcielsForList.map(article => `
  <li class="article">${article.isSecret ? '비밀글입니다.' : article.title}</li>
`).join('');
/*
  <li>title 1</li>
  <li>비밀글입니다</li>
  <li>title 3</li>
*/
```

**Comment by Leeingnyo** [comment]
I didn't know that `Object.valuse()` is ES2017 feature. Thank you for notice. I'll rollback the code using 'Object.keys()'

**Comment by HurSungYun** [comment]
Okay I totally understand what it means and your example is reasonable.

Except Object.value() issue, it's great. Thanks for implementing feature request!

**Pull request merged**